ls1intum/Athena
# Working setup & dependency updates

I did a lot of things to update dependencies & make Athena work on my local machine. Also, I simplified the initial setup down to running one command.

# Changelog

If you think that some of my decisions go in the wrong direction for the project, I'm happy to adapt them!

Items with a ❗ might need changes on the deployment server. The order of the changes follows the file list on GitHub so that it is a bit easier to follow.

- Changed Traefik Dashboard Port to `9081` to avoid conflict with the [default port of a local Jira instance](https://docs.artemis.cit.tum.de/dev/setup/) for Artemis.
- ❗ Renamed the `LOAD_BALANCER_AUTHORIZATION_SECRET` environment variable to `AUTHORIZATION_SECRET`, as it was also used by the clustering service and the name felt wrong because of that.
- ❗ Moved the node configuration files to the load-balancer folder because that's the only service using them. I also applied consistent naming `node_config.<environment>.yml`.
- ❗ Removed the mounted `load-balance.yml` file for the load balancer in `docker-compose.yml` because the file is now contained in the service itself.
- ❗ Introduced `TRACKING_DATABASE_USER` and `TRACKING_DATABASE_PWD` environment variables. Until now, their values were hardcoded in docker-compose, so I don't know how that was managed in production.
- Added new `.local.env` file. It is used when running the microservices without Docker. You can now run all microservices without Docker, which makes it much more convenient to 
    1. use the uvicorn reload feature and
    2. an IDE debugger. 

    You will still need to run traefik and the database using the new file `docker-compose-traefik-db.yml`.
- Added run configurations for PyCharm to simplify the initial setup when using this IDE.
- Added `Makefile`s for everything I needed to do to get Athene microservices running without Docker. You can now run `make setup` in the root directory to 
    - install virtual environments
    - download ML models
    - perform necessary file copy operations and
    - set up protobuf
- Formatted markdown files for codacy to be happy with the PR.
- Upgraded the Python version used in Docker to 3.10, [the latest version that PyTorch currently supports](https://pypi.org/project/torch/1.13.1/).
- Removed piptree files because the piptree project [wasn't updated in three years](https://github.com/mightym/piptree) and I think we should migrate to a more modern approach like [poetry](https://python-poetry.org/) in the future.
- Updated as many packages from PIP as possible. This did not work for the embedding service because [elmo support was dropped from allennlp in v1.0](https://github.com/allenai/allennlp/issues/4384#issuecomment-646724327). Because of this, there are some conflicts between packages which I was only able to resolve by specifying the versions for all dependencies myself and using the pip [`--no-deps` flag](https://stackoverflow.com/questions/12759761/pip-force-install-ignoring-dependencies). Because of this, `embedding/requirements.txt` is not minimal anymore. However, I still ensured the minimality of the other `requirements.txt` files (thanks for the note @jpbernius !)
- Added start scripts to be able to run a Python file to start each microservice instead of `uvicorn`.
- Adapted `docker-compose.override.yml` to
  1. not throw a syntax error (empty service entries caused them for me)
  2. build the protobuf client, so that developers don't have to do that themselves before starting the services.
- Changed the environment variable passing: Now the names are just kept when passing them through. This was necessary to enable a working setup without Docker (where there is no remapping possible). I renamed the used variables within the services themselves as well, so there should be no additional need for configuration change on a production server - except for the already mentioned changes.
- Added a patch for a change in spaCy that broke the embedding service, inspired by https://github.com/allenai/allennlp/pull/4953#issuecomment-774791649 and https://github.com/allenai/allennlp-models/pull/219.
- Changed the error that appears when the model download from Nextcloud fails: It's now a warning. @jpbernius kindly explained to me that this error does not crash the service, so I think that a warning is a better fit here.
- Introduce an environment variable `REWRITE_LOCALHOST_CALLBACK_URL_TO` so that the load-balancer service knows how to call back Artemis when it is running in a Docker container and the Artemis `server_url` is set to `localhost`.

# Test Instructions
1. Set up a local instance of Artemis. Configure Athene exactly [like in the setup instructions](https://docs.artemis.cit.tum.de/dev/setup/#athene-service). Make sure that the Artemis `server_name` in the local configuration is set to `http://localhost:8080` so that Athene knows to call back this endpoint. Also, double-check that the _Athene_ profile is enabled.
2. Clone my fork and check out the `fix/working-setup` branch. It's best to clone from scratch I think: Then there are no old Docker containers / old setup files.
3. There are 2 paths now. In the root directory:
    1. Run `docker-compose up --build`.
    2. Run `make setup -j6` and then `make start`.
4. Go to http://localhost:8080, create a course, then create a text exercise.
5. Use at least 10 test accounts to participate in the exercise. Otherwise, Athene will not start.
6. Edit the exercise settings and set the due date to be the next full minute that has not yet passed.
7. Wait a bit, then log messages should appear in the different microservices in Athene.
8. To verify that Artemis gets called back by Athene, it might be helpful to set a breakpoint in `AtheneService.java` (in Artemis) in the `processResult` method.
